test/selectors-misc.spec.ts

```diff
@@ -75,3 +75,96 @@ it('should work with :visible', async ({page}) => {
 
   expect(await page.$eval('div:visible', div => div.id)).toBe('target2');
 });
+
+it('should work with proximity selectors', async ({page}) => {
+  if (!selectorsV2Enabled())
+    return; // Selectors v1 do not support this.
+
+  /*
+
+       +--+  +--+
+       | 1|  | 2|
+       +--+  ++-++
+       | 3|   | 4|
+  +-------+  ++-++
+  |   0   |  | 5|
+  | +--+  +--+--+
+  | | 6|  | 7|
+  | +--+  +--+
+  |       |
+  O-------+
+          +--+
+          | 8|
+          +--++--+
+              | 9|
+              +--+
+
+  */
+
+  const boxes = [
+    // x, y, width, height
+    [0, 0, 150, 150],
+    [100, 200, 50, 50],
+    [200, 200, 50, 50],
+    [100, 150, 50, 50],
+    [201, 150, 50, 50],
+    [200, 100, 50, 50],
+    [50, 50, 50, 50],
+    [150, 50, 50, 50],
+    [150, -51, 50, 50],
+    [201, -101, 50, 50],
+  ];
+  await page.setContent(`<container style="width: 500px; height: 500px; position: relative;"></container>`);
+  await page.$eval('container', (container, boxes) => {
+    for (let i = 0; i < boxes.length; i++) {
+      const div = document.createElement('div');
+      div.style.position = 'absolute';
+      div.style.overflow = 'hidden';
+      div.style.boxSizing = 'border-box';
+      div.style.border = '1px solid black';
+      div.id = 'id' + i;
+      div.textContent = 'id' + i;
+      const box = boxes[i];
+      div.style.left = box[0] + 'px';
+      // Note that top is a flipped y coordinate.
+      div.style.top = (250 - box[1] - box[3]) + 'px';
+      div.style.width = box[2] + 'px';
+      div.style.height = box[3] + 'px';
+      container.appendChild(div);
+    }
+  }, boxes);
+
+  expect(await page.$eval('div:within(#id0)', e => e.id)).toBe('id6');
+  expect(await page.$eval('div:within(div)', e => e.id)).toBe('id6');
+  expect(await page.$('div:within(#id6)')).toBe(null);
+
+  expect(await page.$eval('div:right-of(#id6)', e => e.id)).toBe('id3');
```

Comment:
can we have `$$eval` here as well?

Will it result in all rects that are to the right?

Reply:
Done.